### PR TITLE
OAuth support for accessing Web API endpoints

### PR DESCRIPTION
**Pull request by kingosticks**
Fixes and tests for @adamcik's original branch.

Still need to consider (from #130):

- [x] check for id/secret from spotify_web, though that would mean making the new ones optional?

### CODE REVIEW

File: mopidy_spotify/web.py [hunk 1]
```diff
@@ -75,7 +75,14 @@ def _refresh_token(self):
...
-        kwargs.setdefault('headers', {}).update(self._headers)
-        return self._request('GET', url, **kwargs)
+        result = {}
+        try:
+            if self._should_refresh_token():
+                self._refresh_token()
+            kwargs.setdefault('headers', {}).update(self._headers)
+            response = self._request('GET', url, **kwargs)
+            result = self._decode(response)
+        except Error as e:
```
Comment (adamcik) [inline mopidy_spotify/web.py]: We should probably catch a more specific error? Also do we want to log this just as debug?

Comment (kingosticks) [inline mopidy_spotify/web.py]: I was thinking could raise and catch `OAuthTokenRefreshError` and a `OAuthClientError` exceptions. And I agree, these want more visibility, error level doesn't seem unreasonable to me.

**Comment by kingosticks** [comment]
I'm done here, any comments on the latest commits? Happy to remove/edit the warning message if anyone has a better idea.

**Comment by adamcik** [comment]
Not sure if caring about the credentials between the versions is the same makes any sense? Other wise this looks fine :-)

**Comment by kingosticks** [comment]
I don't follow, between versions of mopidy-spotify? Which credentials? 

I was thinking maybe it should check the credentials provide the particular scopes we need? Did mopidy-spotify-web previously request whatever scope is needed for search and images or did you have to add that?

**Comment by adamcik** [comment]
As in `spotify.client_* != spotify_web.client_*` probably not really mattering. If that clarifies things a bit.

For the scopes we currently request enough, but as soon as we start moving away from libspotify more whole sale this will start to matter more. And since there are already a bunch of tokens in use we just have to figure out how to migrate people somewhat nicely when the time comes.

**Comment by kingosticks** [comment]
I thought it was necessary for both extensions to use the same set of credentials.  Shouldn't each Spotify user only be using one set of client_id/secret credentials at a time? If not, then I'll just revert that last commit.

**Comment by adamcik** [comment]
You can have as many tokens a you like, and it makes no difference if there are separate ones for the two plugins. What could work is having `spotify` use the one from `spotify_web` if it's own value isn't set. But I think it makes more sense to make the `spotify` one required forcing people to fix this.

**Comment by adamcik** [comment]
Other thing we could do is look over the scopes and reconfigure the defaults the oauthclientbridge requests, potentially reducing the migration required later?

**Comment by kingosticks** [comment]
> You can have as many tokens a you like, and it makes no difference if there are separate ones for the two plugins. 

OK. I'm going to revert that last commit and squash the rest.

> What could work is having spotify use the one from spotify_web if it's own value isn't set. But I think it makes more sense to make the spotify one required forcing people to fix this.

Agreed.

> Other thing we could do is look over the scopes and reconfigure the defaults the oauthclientbridge requests, potentially reducing the migration required later?

I think this would be worthwhile. Getting playlists through the Web API is probably the next most pressing thing to fix so making sure we have all the `playlist-*` scopes would be nice.

**Comment by kingosticks** [comment]
Ready to go! Do we need @jodal to make and upload a release?

**Comment by adamcik** [comment]
I should have access at least, but haven't done this in ages. I'm kinda liking the idea of automating away all of this for everything in the mopidy org, but that is well beyond the scope of this PR. 

I'll go ahead and merge this as step one at least, and I might have time/energy to look at doing a release this evening.

**Comment by pgrenaud** [comment]
Good job on this PR. 👏 I'd love to see a release today! 😄

**Pull request merged**